### ISSUE TITLE
Using UnitOfWork with UnitOfWorkAwareProxyFactory when creating json on the resource does not work anymore

### ISSUE THREAD

**Issue by irineuruiz** [dropwizard/dropwizard#3192]
Hello,

In version 1.3.12 of Dropwizard we used a class that is created with UnitOfWorkAwareProxyFactory.
The return of one of our resources had a variable that is an instance of this class and that during the creation of the response in JSON saved some data in the database via hibernate.

We are migrating to Dropwizard 2.0.2 and we had a problem with that, UnitOfWork tries to use the same transaction as the resource due to the new code implemented
and therefore throws the error due to the code

    if (unitOfWork.transactional ()! = (txn! = null && txn.isActive ())) {
        throw new IllegalStateException (String.format (
            "Existing session transaction state (% s) does not match requested (% b)",
            txn == null? "NULL": Boolean.valueOf (txn.isActive ()),
            unitOfWork.transactional ()
        ));
    }
Because the resource transaction is no longer active and we have the response "Unable to create json".
This worked perfectly in version 1.3.12

I saw that this new code was created at https://github.com/dropwizard/dropwizard/pull/2913

How should I act in such a case?
Is it a bug in the new version?
We are doing something wrong?

**Comment by matthiasblaesing** [comment]
@irineuruiz I'm willing to look at your code - could you provide a minimal, runnable, selfcontained sample, that reproduces your problem?

**Comment by irineuruiz** [comment]
[unitOfWork.zip](https://github.com/dropwizard/dropwizard/files/4331486/unitOfWork.zip)

In this zip I have a simple case, of course this case can be done without the Supplier, but we have a very more complex case in our system...

CREATE TABLE public.tb_person
(
  id integer NOT NULL,
  firstname text,
  CONSTRAINT pk00_tb_person PRIMARY KEY (id)
)
WITH (
  OIDS=FALSE
);

**Comment by matthiasblaesing** [comment]
Thank you - based on this sample I could reproduce the issue. I'm not sure if the problem was just not revealed in the past or if it was introduced by it, but I'm looking into it.

The current summary (I think): Nested UnitOfWorks are supported, interleaved are currently not.

**Comment by matthiasblaesing** [comment]
Correction: These are not interleaved, but serial transactions. This looks like a dangerous setup - my take on this:

@UnitOfWork starts a session and optionally a transaction. The transaction is committed/rolled back on the return from the @UnitOfWork method, the session is kept open. The code reads (UnitOfWorkAspect#afterEnd):

>  // We should not close the session to let the lazy loading work during serializing a response to the client.
>  // If the response successfully serialized, then the session will be closed by the `onFinish` method

That means, that lazy loaded data is initialized in a different/outside a transaction. It also contradicts the documentation (https://www.dropwizard.io/en/latest/manual/hibernate.html):

> The Hibernate session is closed before your resource method’s return value (e.g., the Person from the database), which means your resource method (or DAO) is responsible for initializing all lazily-loaded collections, etc., before returning. Otherwise, you’ll get a LazyInitializationException thrown in your template (or null values produced by Jackson).

@joschi @artem the comment above referenced above was introduced with 7590203f13fb5d728ae6de1946643e4f06361c9a

I agree with the documentation - it is dangerous and badly specified what gets returned if a lazy collection is returned from a different transaction and here we see the fallout.

I have an idea for a fix: Currently the implemented behavior of @UnitOfWork is: If conflicting `transactional` values are found in the @UnitOfWork annotations an Exception is raised. Hypothetical call sequence:

```
(1) UW1-Enter (transactional=true)
(2) UW2-Enter (transactional=false)
(3) UW2-Leave (transactional=false)
(4) UW1-Leave (transactional=true)
```

Currently an exception is raised at point (2).  The problem described here results from this sequence:

```
(1) UW1-Enter (transactional=true)
(2) UW1-Leave (transactional=true)
(3) UW2-Enter (transactional=true)
(4) UW2-Leave (transactional=true)
```

At point (3) an exception is raised - the reason is, that a session is created at point (1) and a transaction is started, at point (2) the transaction is committed. Now the interceptor at point (3) finds the session created at point (1) (which we can't close yet because we want to enable lazy loading), detectects, that is not in an transaction and raises the exception, because  a transaction is required.

If `transactional`=false is interpreted as "can be run without transaction, but can join one" (compareable to "Supports" container managed persistency), we would get:

```
(1) UW1-Enter (transactional=true)     // Transaction T1 created
(2) UW2-Enter (transactional=false)    // T1 still active
(3) UW2-Leave (transactional=false)   // T1 still active
(4) UW1-Leave (transactional=true)    // T1 is commited
```

and for the error case:

```
(1) UW1-Enter (transactional=true)    // Transaction T1 created
(2) UW1-Leave (transactional=true)   // T1 committed
(3) UW2-Enter (transactional=true)    // Transaction T2 created
(4) UW2-Leave (transactional=true)   //  T2 committed
```

So @UnitOfWork(transactional=true) would be compareable to @TransactionAttribute(REQUIRED) and @UnitOfWork(transactional=false) would be compareable to @TransactionAttribute(SUPPORTS)

To complete this - the last relevant sequence would be:

```
(1) UW1-Enter (transactional=false)   
(2) UW2-Enter (transactional=true)    // Transaction T1 created
(3) UW2-Leave (transactional=true)   // T1 is commited
(4) UW1-Leave (transactional=false)
```

Does that sound sane?

**Comment by matthiasblaesing** [comment]
This is the implementation of the suggested handling: https://github.com/matthiasblaesing/dropwizard/commit/082ec3ba5a93aecf330c63f57182314ebc03b995. All existing unittests still succeed, the new unittests failed before this change and with the change succeed.

**Comment by matthiasblaesing** [comment]
@irineuruiz can you comment on the outlined solution and check the proposed patch?

**Comment by irineuruiz** [comment]
I'm sorry, I was on vacation and I just came back today, I'll check it out

**Comment by irineuruiz** [comment]
Hi @matthiasblaesing 

I tested your suggestion and it worked as expected

**Comment by matthiasblaesing** [comment]
Commenting to keep this open. Fix is proposed in #3201 and is pending review.

**Comment by matthiasblaesing** [comment]
The issue is not stale, review is still pending.
